ctlearn/run_model.py

```diff
@@ -439,19 +439,16 @@ def model_fn(features, labels, mode, params):
 
         # Train and evaluate the model
         logger.info("Training and evaluating...")
-        num_validations = config['Training']['num_validations']
-        steps = config['Training']['num_training_steps_per_validation']
-        train_forever = False if num_validations != 0 else True
-        num_validations_remaining = num_validations
-        while train_forever or num_validations_remaining:
-            estimator.train(
-                lambda: input_fn(reader, training_indices, **config['Input']),
-                steps=steps, hooks=hooks)
-            estimator.evaluate(
-                lambda: input_fn(reader, validation_indices, **config['Input']),
-                hooks=hooks, name='validation')
-            if not train_forever:
-                num_validations_remaining -= 1
+        max_steps = config['Training']['max_steps']
+        train_input_fn = lambda: input_fn(reader, training_indices,
+                                          shuffle_and_repeat=True,
+                                          **config['Input'])
+        train_spec = tf.estimator.TrainSpec(input_fn=train_input_fn,
+                                            max_steps=max_steps, hooks=hooks)
+        eval_input_fn = lambda: input_fn(reader, validation_indices,
+                                         **config['Input'])
+        eval_spec = tf.estimator.EvalSpec(input_fn=eval_input_fn)
```

Comment:
I would be in favour of passing the number of training steps per epoch for the `steps` argument, so we evaluate after each epoch. What do you think?

Reply:
The `steps` argument in the `EvalSpec` refers to the number of validation steps, not the number of steps between validations. We want the default (evaluate on the whole validation set). In the `train_and_evaluate` API, the separation between evaluations is based on time, not number of steps. It will evaluate every 2 minutes by default (skipping if there's no new checkpoint). The time between validations is set by the `start_delay_secs` argument. We could make that configurable, though I'm not sure there's a strong need for it.

Reply:
Agree!